From: Gerald Schaefer <geraldsc@de.ibm.com>
Subject: IBM Codestream linux-2.6.16 october2005, patch 02-05

  Description: kernel: Missing error check on signal frame setup.
  Symptom:     User space application does not return to kernel after signal
               handling.
  Problem:     The return value of __put_user() which writes the syscall opcode
               to user space is not checked. Even if this __put_user() fails the
               user space application will run again instead of sending it a
               SIGSEGV.
  Solution:    Check return value of __put_user().
  Problem-ID:  23074
  Archs:       s390-31, s390-64

Acked-by: Hannes Reinecke <hare@suse.de>

--- linux-2.5/arch/s390/kernel/signal.c	20 Mar 2006 15:56:34 -0000	1.26
+++ linux-2.5/arch/s390/kernel/signal.c	10 Apr 2006 10:56:49 -0000	1.26.2.1
@@ -358,8 +358,9 @@
 	} else {
                 regs->gprs[14] = (unsigned long)
 			frame->retcode | PSW_ADDR_AMODE;
-		err |= __put_user(S390_SYSCALL_OPCODE | __NR_rt_sigreturn,
-	                          (u16 __user *)(frame->retcode));
+		if (__put_user(S390_SYSCALL_OPCODE | __NR_rt_sigreturn,
+			       (u16 __user *)(frame->retcode)))
+			goto give_sigsegv;
 	}
 
 	/* Set up backchain. */

